kyo-offheap/jvm/src/main/scala/kyo/Memory.scala

```diff
@@ -0,0 +1,329 @@
+  * all allocations within its scope and automatically frees them when the block completes, whether normally or due to an error.
+  *
+  * IMPORTANT: Care must be taken to ensure Memory objects do not escape their Arena.run scope. While the type system tracks Arena effects,
+  * it is possible to capture a Memory reference and attempt to use it after its Arena has closed. Such usage will result in runtime errors,
+  * as the underlying memory will have been deallocated.
+  *
+  * The implementation is designed with performance in mind, with operations being inlined to avoid boxing and allocation overhead. All
+  * potentially unsafe operations are tracked in the type system through effects, making it explicit which parts of your program perform
+  * memory operations while ensuring they're properly contained within Arena contexts. When maximum performance is required, the unsafe API
+  * allows dropping down to direct memory operations while keeping such operations explicitly marked.
+  *
+  * Note that memory allocation operations (init, initWith) are intentionally not provided in the Unsafe API. This ensures that all memory
+  * allocations are tracked by the Arena effect, which is necessary for automatic resource cleanup.
+  *
+  * @tparam A
+  *   The type of elements stored in this memory segment
+  */
+opaque type Memory[A] = MemorySegment
+
+export Memory.Arena
+
+object Memory:
+
+    given [A]: Flat[Memory[A]] = Flat.derive
+
+    /** Initializes a new memory segment of the specified size for type A.
+      *
+      * @tparam A
+      *   The type of elements to store
+      * @param size
+      *   The number of elements the memory segment can hold
+      * @return
+      *   A new Memory instance within an Arena context
+      */
+    def init[A: Layout as l](size: Int)(using Frame): Memory[A] < Arena =
+        initWith[A](size)(identity)
+
+    /** Initializes a memory segment and immediately applies a function to it.
+      *
+      * @tparam A
+      *   The type of elements to store
+      * @param size
+      *   The number of elements the memory segment can hold
+      * @param f
+      *   The function to apply to the newly created memory
+      * @return
+      *   The result of applying f to the new memory segment
+      */
+    inline def initWith[A: Layout as l](size: Int)[B, S](inline f: Memory[A] => B < S)(using inline frame: Frame): B < (Arena & S) =
+        Arena.use { arena =>
+            IO.Unsafe(f(arena.allocate(l.size * size)))
+        }
+
+    extension [A: Layout as l](self: Memory[A])
+
+        /** Retrieves the element at the specified index.
+          *
+          * @param index
+          *   The position to read from
+          * @return
+          *   The element at the specified index
+          */
+        inline def get(index: Int)(using inline frame: Frame): A < Arena =
+            IO.Unsafe {
+                val x = Unsafe.get(self)(index)
+                x
+            }
```

Comment:
Is this leftover/inference issue?
```suggestion
            IO.Unsafe(Unsafe.get(self)(index))
```

Reply:
ah, I forgot to mention this. Yes, I get the missing `AllowUnsafe` compilation failure without the `val`. I've added a comment about it